### PR TITLE
Update the Build from source documentation

### LINKED ISSUE #16790: Update the Build from source documentation
### Description

Hello,

Since version 20 the build from source documentation is not up-to-date

https://github.com/keycloak/keycloak/blob/main/docs/building.md

For exemple I can't build only the server anymore

The server-dist folder doesn't exist anymore

`mvn -Pdistribution -pl distribution/server-dist -am -Dmaven.test.skip clean install`

What is the new way to achieve the same goal ? Thanks.

### PR DESCRIPTION
**Pull request by abstractj**
Closes #16790

Maybe we should consider merging this document with https://github.com/keycloak/keycloak/blob/main/quarkus/README.md now that the WildFly dist was deprecated.

### CODE REVIEW

**Review by mposolda — CHANGES_REQUESTED** [review]
I still see some issue with this:

1) In the initial "Bulding from source" section, there is still the line stating: Once completed you will find distribution archives in `distribution`.
 
This one is not very accurate as the most important distribution archive is the server inside directory `quarkus/dist/target/`. AFAIK the `distribution` directory still contains some pieces (EG. adapters etc), but not very important for typical developer. 

2) This is not a blocker for this PR, but rather usability issue. Should we change the commands `mvn clean install` and `mvn clean install -Pdistribution` in the initial "Building from source" section to automatically contain `-DskipTests=true` ?

The whole testsuite runs around 3 hours and there is not much point in run it on your laptop since we have GH actions anyway. We can instead add the additional paragraph below about how to run testsuite.

EG. add something like: If you want to run testsuite, use command `mvn clean install`, but keep in mind that testsuite runs few hours to complete. If you want to check if your change caused regression in the testsuite, it may be more effective to push your changes to your github fork and then look at github actions build if the tests were successful. You can find the build when looking for example at https://github.com/your-github-username/keycloak/tree/your-github-branch and then click the button right before your commit ID.

WDYT?

**Comment by abstractj** [comment]
@mposolda absolutely np about introducing those changes Marek. For us to collaborate and others to commented, I created the following document https://hackmd.io/@abstractj/SkTH4szai. You should have write access to that, so please feel free to include, or comment on, anything that might be missing.

As soon as we are happy with the content, I can update the changes here.

**Comment by mposolda** [comment]
@abstractj Nice tool! Added some suggestions inline.

**Comment by abstractj** [comment]
@mposolda thank you, introduced the suggestions you made and included you as co-author on my commit.

File: docs/building.md [hunk 1]
```diff
@@ -30,19 +30,21 @@ First clone the Keycloak repository:
...
+    mvn clean install -DskipTests=true
     
-This will build all modules and run the testsuite. 
+This will build all modules. The build of the Keycloak server is then available in the archive under the directory `quarkus/dist/target`.
 
-To build the ZIP distribution run:
+If you would also like to run testsuite during the build, you can omit the `-DskipTests=true` parameter. Please be aware that testsuite can take at least 3 hours to complete. However, there is an alternative if you want to avoid running the full testsuite on your laptop and at the same time ensure that your changes in the code did not cause regressions. To check if your changes will cause a regression, you should push the changes to a GitHub fork and verify if all the checks are successful. Those checks run testsuite and do other verifications.
 
-    mvn clean install -Pdistribution
+To build the ZIP for the WildFly distribution (*deprecated*) with all the adapters:
```
Comment (stianst) [inline docs/building.md:39]: There is no longer a WF distribution in main

Comment (abstractj) [reply docs/building.md]: Thanks @stianst, updated.

**Comment by Cracky5457** [comment]
I don't want to be off-topic here but it's a bit about documentation,

It would be nice to improve the `Working with the codebase` section to explain a bit more how to contribute and work on PR, I was not able to simply execute keycloak in debug mode with Eclipse or IntelliJ since keycloak 20 and the switch to quarkus and it's a pain.

I've been creating a Q&A here, some people seems interested too. It would be very nice to enhance contribution on keycloak if you can take few minutes for that :)

https://github.com/keycloak/keycloak/discussions/16791

Thanks

**Comment by Cracky5457** [comment]
https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/HOW-TO-RUN.md

Maybe this should also be adapted for quarkus ?

File: docs/building.md [hunk 2]
```diff
@@ -30,19 +30,17 @@ First clone the Keycloak repository:
...
 To build Keycloak run:
 
-    mvn clean install
+    mvn clean install -DskipTests=true
     
-This will build all modules and run the testsuite. 
+This will build all modules. The build of the Keycloak server is then available in the archive under the directory `quarkus/dist/target`.
 
-To build the ZIP distribution run:
+If you would also like to run testsuite during the build, you can omit the `-DskipTests=true` parameter. Please be aware that testsuite can take at least 3 hours to complete. However, there is an alternative if you want to avoid running the full testsuite on your laptop and at the same time ensure that your changes in the code did not cause regressions. To check if your changes will cause a regression, you should push the changes to a GitHub fork and verify if all the checks are successful. Those checks run testsuite and do other verifications.
```
Comment (stianst) [inline docs/building.md:37]: I'd probably leave this out, as it should rather be captured in the testing documentation

File: docs/building.md [hunk 3]
```diff
@@ -30,19 +30,17 @@ First clone the Keycloak repository:
     
 To build Keycloak run:
 
-    mvn clean install
+    mvn clean install -DskipTests=true
```
Comment (stianst) [inline docs/building.md:33]: Should we also add?
```suggestion
    mvn clean install -DskipTests=true -DskipTestsuite
```

File: docs/building.md [hunk 4]
```diff
@@ -1,14 +1,14 @@
...
 
     java -version
     mvn -version
     git --version
 
 **NOTE**
 
-If you use Maven 3.8.1 or newer, please add the following mirror to your local
+If you are using Maven 3.8.1 or newer, please add the following mirror to your local 
 `~/.m2/settings.xml` to avoid build errors:
```
Comment (stianst) [inline docs/building.md:12]: This is not needed anymore

File: docs/building.md [hunk 5]
```diff
@@ -1,14 +1,14 @@
 ## Building from source
 
-Ensure you have JDK 11 (or newer), Maven 3.5.4 (or newer) and Git installed
+Make sure you have JDK 11 (or newer), Maven 3.5.4 (or newer) and Git installed.
```
Comment (stianst) [inline docs/building.md:3]:
```suggestion
Make sure you have JDK 11 (or newer), Maven 3.8.6 (or newer) and Git installed.
```

Maybe we could switch the instructions here to use `./mvnw`?

File: docs/building.md (hunk 2 above)
Comment (abstractj) [reply docs/building.md:37]: Those were the changes suggested by @mposolda. I don't mind removing it if we agree.

**Comment by stianst** [comment]
Closing as this was resolved as part of https://github.com/keycloak/keycloak/commit/08838f77caad92d0ef7bd320494dce83ec034427

**Pull request closed without merging**